### PR TITLE
rpc: add chunked rpc interface

### LINKED ISSUE #6354: rpc: genesis endpoint
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): Present on master.

**ABCI app** (name for built-in, URL for self-written if it's publicly available): Cosmos-sdk

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:

When querying the Tendermint RPC I was not able to receive the entire genesis due to the network request timing out.

**What you expected to happen**:

Receive the entire genesis file

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):

Visit https://rpc.cosmos.network/genesis?

**Proposed solution**

Introduce a new endpoint `genesis_chunk` which can be used to receive a specific chunk. The total amount of chunks will be sent within the response as well. 

The response will be:

```
{
  "chunk": 1,
  "total": 15,
  "data: <base64-blob>
}
```

### PR DESCRIPTION
**Pull request by tychoish**
I think there may be additional testing that would be useful, and I
think we should also update documentation and specs, maybe, but I'm
not sure the full scope or ordering of that work relative to this PR.

Resolves #6354

### CODE REVIEW

File: light/proxy/routes.go [hunk 1]
```diff
@@ -92,6 +92,14 @@ func makeGenesisFunc(c *lrpc.Client) rpcGenesisFunc {
 	}
 }
 
+type rpcGenesisChunkedFunc func(ctx *rpctypes.Context, chunkID uint) (*ctypes.ResultGenesisChunk, error)
+
+func makeGenesisChunkedFunc(c *lrpc.Client) rpcGenesisChunkedFunc {
```
Comment (cmwaters) [inline light/proxy/routes.go:97]: I think you need to add this to the light client routes further up this page (like you did for `rpc/core/routers.go`

File: rpc/client/rpc_test.go [hunk 2]
```diff
@@ -186,6 +188,31 @@ func TestGenesisAndValidators(t *testing.T) {
...
+func TestGenesisChunked(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for _, c := range GetClients() {
+		first, err := c.GenesisChunked(ctx, 1)
+		require.NoError(t, err)
+
+		decoded := make([]string, 0, first.TotalChunks)
+		for i := first.ChunkNumber; i < first.TotalChunks+1; i++ {
```
Comment (cmwaters) [inline rpc/client/rpc_test.go:200]: How big is this test genesis file? Does it even loop more than once?

File: rpc/client/rpc_test.go [hunk 3]
```diff
@@ -186,6 +188,31 @@ func TestGenesisAndValidators(t *testing.T) {
...
+			require.NoError(t, err)
+			data, err := base64.StdEncoding.DecodeString(chunk.Data)
+			require.NoError(t, err)
+			decoded = append(decoded, string(data))
+
+		}
+		doc := []byte(strings.Join(decoded, ""))
+
+		var out types.GenesisDoc
+		require.NoError(t, tmjson.Unmarshal(doc, &out),
```
Comment (cmwaters) [inline rpc/client/rpc_test.go:211]: Being able to unmarshal basically ensures that the chunked genesis is the same as the one before right? Although technically they could be different

**Review by cmwaters — COMMENTED** [review]
This all looks good to me. I think the first thing you should do is open a PR in the spec repo looking to add this rpc endpoint. Then we can merge this one / add more tests if you feel like some edge cases are missing. After that we can update the docs - which I think is mainly updating the swagger.yml file.

**Review by alexanderbez — APPROVED** [review]
Looks great @tychoish. I just a few comments 👍

**Comment by tychoish** [comment]
@cmwaters (et al) I added this to the spec: https://github.com/tendermint/spec/pull/299

File: rpc/client/rpc_test.go (hunk 2 above)
Comment (cmwaters) [reply rpc/client/rpc_test.go:207]: Do we want to create a test where there are more than one chunk?

File: rpc/core/env.go [hunk 4]
```diff
@@ -25,6 +27,9 @@ const (
 	// SubscribeTimeout is the maximum time we wait to subscribe for an event.
 	// must be less than the server's write timeout (see rpcserver.DefaultConfig)
 	SubscribeTimeout = 5 * time.Second
+
+	// genesisChunkSize is the size, in bytes of each chunk in
```
Comment (cmwaters) [inline rpc/core/env.go:31]:
```suggestion
	// genesisChunkSize is the maximum size, in bytes, of each chunk
```

File: rpc/core/net.go [hunk 5]
```diff
@@ -95,6 +95,20 @@ func (env *Environment) Genesis(ctx *rpctypes.Context) (*ctypes.ResultGenesis, e
 	return &ctypes.ResultGenesis{Genesis: env.GenDoc}, nil
 }
 
+func (env *Environment) GenesisChunked(ctx *rpctypes.Context, chunkID uint) (*ctypes.ResultGenesisChunk, error) {
+	id := int(chunkID)
```
Comment (cmwaters) [inline rpc/core/net.go:99]: Do we want to add a check here that we have initialised the chunks. Will this fail acceptably if we haven't initialised the chunks?

**Review by cmwaters — APPROVED** [review]

File: rpc/core/net.go (hunk 5 above)
Comment (tychoish) [reply rpc/core/net.go]: I added the check that if we give an ID that is out of bounds we'll generate an error, so it's "safe".

**Review by marbar3778 — APPROVED** [review]
utACK. 

Could you add a changelog and upgrading.md entry.

**Comment by alexanderbez** [comment]
@tychoish what's stopping this from being merged?

**Comment by marbar3778** [comment]
@mergifyio backport release/v0.34.x

**Comment by marbar3778** [comment]
@Mergifyio backport v0.34.x

**Comment by cmwaters** [comment]
I don't see the `/genesis_chunked` endpoint in the RPC documentation (https://docs.tendermint.com/master/rpc/). Is there something we're missing?

**Comment by marbar3778** [comment]
nope, it wasn't added in the original PR. 

will make an issue
